core/src/apps/cardano/auxiliary_data.py

```diff
@@ -0,0 +1,252 @@
+from trezor.crypto import hashlib
+from trezor.crypto.curve import ed25519
+from trezor.messages import (
+    CardanoAddressType,
+    CardanoAuxiliaryDataType,
+    CardanoMetadataType,
+)
+
+from apps.common import cbor
+
+from ..common.seed import remove_ed25519_prefix
+from .address import derive_address_bytes, derive_human_readable_address
+from .helpers import INVALID_AUXILIARY_DATA, bech32
+from .helpers.bech32 import HRP_JORMUN_PUBLIC_KEY
+from .helpers.paths import SCHEMA_STAKING_ANY_ACCOUNT
+from .layout import confirm_catalyst_registration, show_auxiliary_data_hash
+
+if False:
+    from typing import Union
+    from trezor import wire
+
+    from trezor.messages.CardanoCatalystRegistrationParametersType import (
+        CardanoCatalystRegistrationParametersType,
+    )
+    from trezor.messages.CardanoTxAuxiliaryDataType import CardanoTxAuxiliaryDataType
+
+    CatalystRegistrationPayload = dict[int, Union[bytes, int]]
+    CatalystRegistrationSignature = dict[int, bytes]
+    CatalystRegistration = dict[
+        int, Union[CatalystRegistrationPayload, CatalystRegistrationSignature]
+    ]
+
+    from . import seed
+
+AUXILIARY_DATA_HASH_SIZE = 32
+PUBLIC_KEY_LENGTH = 32
+CATALYST_REGISTRATION_HASH_SIZE = 32
+CATALYST_REGISTRATION_KEY = 61284
+CATALYST_REGISTRATION_SIGNATURE_KEY = 61285
+
+
+def validate_auxiliary_data(
+    auxiliary_data: CardanoTxAuxiliaryDataType | None,
+) -> None:
+    if not auxiliary_data:
+        return
+
+    if auxiliary_data.type == CardanoAuxiliaryDataType.BLOB:
+        if not auxiliary_data.blob:
+            raise INVALID_AUXILIARY_DATA
+
+        _validate_auxiliary_data_blob(auxiliary_data.blob)
+    elif auxiliary_data.type == CardanoAuxiliaryDataType.TUPLE:
+        # currently we only support CATALYST_REGISTRATION
+        if (
+            auxiliary_data.blob
+            or not auxiliary_data.metadata
+            or auxiliary_data.metadata.type != CardanoMetadataType.CATALYST_REGISTRATION
+        ):
+            raise INVALID_AUXILIARY_DATA
```

Comment:
Makes sense. I've split split out metadata validation completely so that `validate_auxiliary_data` operates on a single level and can later be more easily extended with auxiliary scripts (or other auxiliary things). I've done the same for `show_auxiliary_data` and `get_auxiliary_data_cbor` so it would be consistent. https://github.com/vacuumlabs/trezor-firmware/pull/57/commits/de6ce66e59dc846769a71ec5ce9fa3e0583d45a1#diff-449dbb3cc07455194600fd241faad462fb0b7417008b0b99fef57c4c7966745bR56